File: pascal/src/parsepattern.pas
unit ParsePattern;

{$mode objfpc}{$H+}

interface

uses
  Token;

type
  TStringArray = array of string;

  TDictionaries = class
  private
    FNames: TStringArray;
    FWords: array of TStringArray;
  public
    procedure SetWords(const Name: string; Words: TStringArray);
    function Has(const Name: string): Boolean;
    function Get(const Name: string): TStringArray;
    function Names: TStringArray;
    function Count: Integer;
  end;

  TTokenArray = array of TToken;

function ParsePattern(const InputPattern: string; Dictionaries: TDictionaries): TTokenArray;
function CharsAsVariants(const S: string): TStringArray;

implementation

uses
  SysUtils;

function TDictionaries.Count: Integer;
begin
  Result := Length(FNames);
end;

function TDictionaries.Names: TStringArray;
begin
  Result := FNames;
end;

procedure TDictionaries.SetWords(const Name: string; Words: TStringArray);
var
  I: Integer;
begin
  for I := 0 to High(FNames) do
    if FNames[I] = Name then
    begin
      FWords[I] := Words;
      Exit;
    end;
  SetLength(FNames, Length(FNames) + 1);
  SetLength(FWords, Length(FWords) + 1);
  FNames[High(FNames)] := Name;
  FWords[High(FWords)] := Words;
end;

function TDictionaries.Has(const Name: string): Boolean;
var
  I: Integer;
begin
  Result := False;
  for I := 0 to High(FNames) do
    if FNames[I] = Name then
      Exit(True);
end;

function TDictionaries.Get(const Name: string): TStringArray;
var
  I: Integer;
begin
  SetLength(Result, 0);
  for I := 0 to High(FNames) do
    if FNames[I] = Name then
      Exit(FWords[I]);
end;

function CharsAsVariants(const S: string): TStringArray;
var
  I: Integer;
begin
  SetLength(Result, Length(S));
  for I := 1 to Length(S) do
    Result[I - 1] := S[I];
end;

function IsSpecial(C: Char): Boolean;
begin
  Result := Pos(C, '#@$*&?!-%') > 0;
end;

function SplitKeepingDelimiters(const Input: string): TStringArray;
var
  I, J, LiteralStart, Len: Integer;
  C: Char;
  Parts: TStringArray;

  procedure PushPart(const Part: string);
  begin
    if Part = '' then
      Exit;
    SetLength(Parts, Length(Parts) + 1);
    Parts[High(Parts)] := Part;
  end;

begin
  SetLength(Parts, 0);
  if Input = '' then
  begin
    Result := Parts;
    Exit;
  end;

  I := 1;
  LiteralStart := 1;
  Len := Length(Input);

  while I <= Len do
  begin
    C := Input[I];
    if (C = '\') and (I + 1 <= Len) and IsSpecial(Input[I + 1]) then
    begin
      if I > LiteralStart then
        PushPart(Copy(Input, LiteralStart, I - LiteralStart));
      PushPart(Copy(Input, I, 2));
      I := I + 2;
      LiteralStart := I;
    end
    else if IsSpecial(C) and (I + 1 <= Len) and (Input[I + 1] = '{') then
    begin
      if I > LiteralStart then
        PushPart(Copy(Input, LiteralStart, I - LiteralStart));
      J := I + 2;
      while (J <= Len) and (Input[J] <> '}') do
        Inc(J);
      if (J <= Len) and (Input[J] = '}') then
      begin
        PushPart(Copy(Input, I, J - I + 1));
        I := J + 1;
        LiteralStart := I;
      end
      else
      begin
        if I > LiteralStart then
          PushPart(Copy(Input, LiteralStart, I - LiteralStart));
        PushPart(C);
        Inc(I);
        LiteralStart := I;
      end;
    end
    else if IsSpecial(C) then
    begin
      if I > LiteralStart then
        PushPart(Copy(Input, LiteralStart, I - LiteralStart));
      PushPart(C);
      Inc(I);
      LiteralStart := I;
    end
    else
      Inc(I);
  end;

  if LiteralStart <= Len then
    PushPart(Copy(Input, LiteralStart, Len - LiteralStart + 1));

  Result := Parts;
end;

function ParseLengthWithVariants(const Part: string; out StartLength, EndLength: Integer): Boolean;
var
  OpenPos, ClosePos, DashPos: Integer;
  Inner, Left, Right: string;
  S, E, N: Integer;
begin
  StartLength := 1;
  EndLength := 1;
  Result := True;
  OpenPos := Pos('{', Part);
  if OpenPos = 0 then
    Exit;
  ClosePos := Pos('}', Part);
  if (ClosePos = 0) or (ClosePos < OpenPos) then
    Exit;
  Inner := Copy(Part, OpenPos + 1, ClosePos - OpenPos - 1);
  DashPos := Pos('-', Inner);
  if DashPos > 0 then
  begin
    Left := Copy(Inner, 1, DashPos - 1);
    Right := Copy(Inner, DashPos + 1, MaxInt);
    if TryStrToInt(Left, S) and TryStrToInt(Right, E) then
    begin
      StartLength := S;
      EndLength := E;
    end;
  end
  else if TryStrToInt(Inner, N) then
  begin
    StartLength := N;
    EndLength := N;
  end;
end;

function ParseLengthWithString(const Part: string; out Content: string;
  out StartLength, EndLength: Integer): Boolean;
var
  OpenPos, AfterOpen, CloseQuote, I: Integer;
  Rest, AfterQuote, Stripped, Left, Right: string;
  DashPos: Integer;
  S, E, N: Integer;
begin
  Result := False;
  StartLength := 1;
  EndLength := 1;
  Content := '';

  OpenPos := Pos('{''', Part);
  if OpenPos = 0 then
    Exit;

  AfterOpen := OpenPos + 2;
  Rest := Copy(Part, AfterOpen, MaxInt);
  CloseQuote := 0;
  for I := Length(Rest) downto 1 do
    if Rest[I] = '''' then
    begin
      CloseQuote := I;
      Break;
    end;
  if CloseQuote = 0 then
    Exit;

  Content := Copy(Rest, 1, CloseQuote - 1);
  AfterQuote := Copy(Rest, CloseQuote + 1, MaxInt);

  if (Length(AfterQuote) = 0) or ((AfterQuote[1] <> '}') and (AfterQuote[1] <> ',')) then
  begin
    if Pos('}', AfterQuote) = 0 then
      Exit;
  end;

  if (Length(AfterQuote) > 0) and (AfterQuote[1] = ',') then
  begin
    Stripped := Copy(AfterQuote, 2, MaxInt);
    if (Length(Stripped) > 0) and (Stripped[Length(Stripped)] = '}') then
      SetLength(Stripped, Length(Stripped) - 1);
    DashPos := Pos('-', Stripped);
    if DashPos > 0 then
    begin
      Left := Copy(Stripped, 1, DashPos - 1);
      Right := Copy(Stripped, DashPos + 1, MaxInt);
      if TryStrToInt(Left, S) and TryStrToInt(Right, E) then
      begin
        StartLength := S;
        EndLength := E;
      end;
    end
    else if TryStrToInt(Stripped, N) then
    begin
      StartLength := N;
      EndLength := N;
    end;
  end
  else if (Length(AfterQuote) = 0) or (AfterQuote[1] <> '}') then
    Exit;

  Result := True;
end;

function MakeLiteralToken(const Part: string): TToken;
var
  Variants: array[0..0] of string;
begin
  Variants[0] := Part;
  Result := CreateTokenDefaults(Part, Variants);
end;

function SimpleTokenizer(const Part, Alphabet: string): TToken;
var
  StartLength, EndLength: Integer;
  Variants: TStringArray;
begin
  Variants := CharsAsVariants(Alphabet);
  ParseLengthWithVariants(Part, StartLength, EndLength);
  Result := CreateToken(Part, StartLength, EndLength, Variants);
end;

function DictionaryTokenizer(const Part: string; Dictionaries: TDictionaries): TToken;
var
  Content: string;
  StartLength, EndLength: Integer;
  Variants: TStringArray;
begin
  if (not ParseLengthWithString(Part, Content, StartLength, EndLength))
    or ((Content <> '') and (not Dictionaries.Has(Content))) then
    Exit(MakeLiteralToken(Part));

  Variants := Dictionaries.Get(Content);
  Result := CreateToken(Part, StartLength, EndLength, Variants);
end;

function UnescapeCommas(const S: string): string;
var
  I: Integer;
begin
  Result := '';
  I := 1;
  while I <= Length(S) do
  begin
    if (I < Length(S)) and (S[I] = '\') and (S[I + 1] = ',') then
    begin
      Result := Result + ',';
      Inc(I, 2);
    end
    else
    begin
      Result := Result + S[I];
      Inc(I);
    end;
  end;
end;

function WordsTokenizer(const Part: string): TToken;
var
  Content, WorkString: string;
  StartLength, EndLength, Index: Integer;
  Variants: TStringArray;
begin
  if not ParseLengthWithString(Part, Content, StartLength, EndLength) then
    Exit(MakeLiteralToken(Part));

  SetLength(Variants, 0);
  WorkString := Content;
  Index := 1;
  while Index <= Length(WorkString) do
  begin
    if (Index < Length(WorkString)) and (WorkString[Index] = '\')
      and (WorkString[Index + 1] = ',') then
      Inc(Index, 2)
    else if WorkString[Index] = ',' then
    begin
      SetLength(Variants, Length(Variants) + 1);
      Variants[High(Variants)] := Copy(WorkString, 1, Index - 1);
      WorkString := Copy(WorkString, Index + 1, MaxInt);
      Index := 1;
    end
    else
      Inc(Index);
  end;
  SetLength(Variants, Length(Variants) + 1);
  Variants[High(Variants)] := WorkString;

  for Index := 0 to High(Variants) do
    Variants[Index] := UnescapeCommas(Variants[Index]);

  Result := CreateToken(Part, StartLength, EndLength, Variants);
end;

function PartToToken(const Part: string; Dictionaries: TDictionaries): TToken;
var
  Variants: array[0..0] of string;
  First: Char;
begin
  if Part = '' then
    Exit(MakeLiteralToken(Part));

  First := Part[1];
  case First of
    '#': Exit(SimpleTokenizer(Part, '0123456789'));
    '@': Exit(SimpleTokenizer(Part, 'abcdefghijklmnopqrstuvwxyz'));
    '*': Exit(SimpleTokenizer(Part, 'abcdefghijklmnopqrstuvwxyz0123456789'));
    '-': Exit(SimpleTokenizer(Part,
        'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'));
    '!': Exit(SimpleTokenizer(Part, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'));
    '?': Exit(SimpleTokenizer(Part, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'));
    '&': Exit(SimpleTokenizer(Part,
        'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'));
    '%': Exit(DictionaryTokenizer(Part, Dictionaries));
    '$': Exit(WordsTokenizer(Part));
  end;

  if (Length(Part) > 1) and (Part[1] = '\') and IsSpecial(Part[2]) then
  begin
    Variants[0] := Copy(Part, 2, MaxInt);
    Exit(CreateTokenDefaults(Part, Variants));
  end;

  Result := MakeLiteralToken(Part);
end;

function ParsePattern(const InputPattern: string; Dictionaries: TDictionaries): TTokenArray;
var
  Parts: TStringArray;
  I: Integer;
  Dicts: TDictionaries;
begin
  SetLength(Result, 0);
  if Dictionaries = nil then
  begin
    Dicts := TDictionaries.Create;
    try
      Result := ParsePattern(InputPattern, Dicts);
    finally
      Dicts.Free;
    end;
    Exit;
  end;

  Parts := SplitKeepingDelimiters(InputPattern);
  for I := 0 to High(Parts) do
    if Parts[I] <> '' then
    begin
      SetLength(Result, Length(Result) + 1);
      Result[High(Result)] := PartToToken(Parts[I], Dictionaries);
    end;
end;

end.

File: fortran/src/parse_pattern.f90
! Hand-rolled pattern split and tokenizers (no regex).
module wl_parse_pattern
  use iso_fortran_env, only: int64
  use wl_util
  use wl_token
  implicit none
  private

  public :: token_list, token_list_free, parse_pattern
  public :: chars_as_variants

  type :: token_list
    type(token_t), allocatable :: items(:)
    integer :: n = 0
  end type token_list

contains

  subroutine token_list_free(list)
    type(token_list), intent(inout) :: list
    integer :: i
    if (allocated(list%items)) then
      do i = 1, list%n
        call token_free(list%items(i))
      end do
      deallocate(list%items)
    end if
    list%n = 0
  end subroutine token_list_free

  subroutine token_list_push(list, tok)
    type(token_list), intent(inout) :: list
    type(token_t), intent(inout) :: tok
    type(token_t), allocatable :: tmp(:)
    integer :: new_cap, i, old_cap
    old_cap = 0
    if (allocated(list%items)) old_cap = size(list%items)
    if (list%n >= old_cap) then
      if (old_cap == 0) then
        new_cap = 4
      else
        new_cap = old_cap * 2
      end if
      allocate(tmp(new_cap))
      do i = 1, list%n
        tmp(i) = list%items(i)
        ! prevent double-free of moved components
        call str_list_init(list%items(i)%variants)
        if (allocated(list%items(i)%src)) deallocate(list%items(i)%src)
      end do
      call move_alloc(tmp, list%items)
    end if
    list%n = list%n + 1
    list%items(list%n) = tok
    call str_list_init(tok%variants)
    if (allocated(tok%src)) deallocate(tok%src)
  end subroutine token_list_push

  subroutine chars_as_variants(s, out)
    character(len=*), intent(in) :: s
    type(str_list), intent(out) :: out
    integer :: i
    call str_list_init(out)
    do i = 1, len(s)
      call str_list_push(out, s(i:i))
    end do
  end subroutine chars_as_variants

  pure logical function is_special(c) result(ok)
    character(len=1), intent(in) :: c
    ok = (index('#@$*&?!-%', c) > 0)
  end function is_special

  subroutine split_keeping_delimiters(input, parts)
    character(len=*), intent(in) :: input
    type(str_list), intent(out) :: parts
    integer :: i, j, literal_start, n
    character(len=1) :: c

    call str_list_init(parts)
    n = len(input)
    if (n == 0) return

    i = 1
    literal_start = 1
    do while (i <= n)
      c = input(i:i)
      if (c == '\' .and. i + 1 <= n .and. is_special(input(i+1:i+1))) then
        if (i > literal_start) call str_list_push(parts, input(literal_start:i-1))
        call str_list_push(parts, input(i:i+1))
        i = i + 2
        literal_start = i
      else if (is_special(c) .and. i + 1 <= n .and. input(i+1:i+1) == '{') then
        if (i > literal_start) call str_list_push(parts, input(literal_start:i-1))
        j = i + 2
        do while (j <= n .and. input(j:j) /= '}')
          j = j + 1
        end do
        if (j <= n .and. input(j:j) == '}') then
          call str_list_push(parts, input(i:j))
          i = j + 1
          literal_start = i
        else
          if (i > literal_start) call str_list_push(parts, input(literal_start:i-1))
          call str_list_push(parts, c)
          i = i + 1
          literal_start = i
        end if
      else if (is_special(c)) then
        if (i > literal_start) call str_list_push(parts, input(literal_start:i-1))
        call str_list_push(parts, c)
        i = i + 1
        literal_start = i
      else
        i = i + 1
      end if
    end do
    if (literal_start <= n) call str_list_push(parts, input(literal_start:n))
  end subroutine split_keeping_delimiters

  subroutine parse_length_with_variants(part, start_length, end_length)
    character(len=*), intent(in) :: part
    integer, intent(out) :: start_length, end_length
    integer :: open_pos, close_pos, dash_pos, s, e, nn
    character(len=:), allocatable :: inner, left, right

    start_length = 1
    end_length = 1
    open_pos = index(part, '{')
    if (open_pos == 0) return
    close_pos = index(part, '}')
    if (close_pos == 0 .or. close_pos < open_pos) return
    inner = part(open_pos+1:close_pos-1)
    dash_pos = index(inner, '-')
    if (dash_pos > 0) then
      left = inner(1:dash_pos-1)
      right = inner(dash_pos+1:)
      if (try_parse_int(left, s) .and. try_parse_int(right, e)) then
        start_length = s
        end_length = e
      end if
    else if (try_parse_int(inner, nn)) then
      start_length = nn
      end_length = nn
    end if
  end subroutine parse_length_with_variants

  logical function parse_length_with_string(part, content, start_length, end_length) result(ok)
    character(len=*), intent(in) :: part
    character(len=:), allocatable, intent(out) :: content
    integer, intent(out) :: start_length, end_length
    integer :: open_pos, after_open, close_quote, i, dash_pos, s, e, nn
    character(len=:), allocatable :: rest, after_quote, stripped, left, right

    ok = .false.
    start_length = 1
    end_length = 1
    content = ''

    open_pos = index(part, "{'")
    if (open_pos == 0) return

    after_open = open_pos + 2
    rest = part(after_open:)
    close_quote = 0
    do i = len(rest), 1, -1
      if (rest(i:i) == "'") then
        close_quote = i
        exit
      end if
    end do
    if (close_quote == 0) return

    content = rest(1:close_quote-1)
    after_quote = rest(close_quote+1:)

    if (len(after_quote) == 0) then
      return
    end if
    if (after_quote(1:1) /= '}' .and. after_quote(1:1) /= ',') then
      if (index(after_quote, '}') == 0) return
    end if

    if (after_quote(1:1) == ',') then
      stripped = after_quote(2:)
      if (len(stripped) > 0 .and. stripped(len(stripped):len(stripped)) == '}') then
        stripped = stripped(1:len(stripped)-1)
      end if
      dash_pos = index(stripped, '-')
      if (dash_pos > 0) then
        left = stripped(1:dash_pos-1)
        right = stripped(dash_pos+1:)
        if (try_parse_int(left, s) .and. try_parse_int(right, e)) then
          start_length = s
          end_length = e
        end if
      else if (try_parse_int(stripped, nn)) then
        start_length = nn
        end_length = nn
      end if
    else if (after_quote(1:1) /= '}') then
      return
    end if

    ok = .true.
  end function parse_length_with_string

  subroutine make_literal_token(part, tok)
    character(len=*), intent(in) :: part
    type(token_t), intent(out) :: tok
    type(str_list) :: variants
    call str_list_init(variants)
    call str_list_push(variants, part)
    call token_init(tok, part, 1, 1, variants)
  end subroutine make_literal_token

  subroutine simple_tokenizer(part, alphabet, tok)
    character(len=*), intent(in) :: part, alphabet
    type(token_t), intent(out) :: tok
    type(str_list) :: variants
    integer :: start_length, end_length
    call chars_as_variants(alphabet, variants)
    call parse_length_with_variants(part, start_length, end_length)
    call token_init(tok, part, start_length, end_length, variants)
  end subroutine simple_tokenizer

  subroutine dictionary_tokenizer(part, dicts, tok)
    character(len=*), intent(in) :: part
    type(dictionaries), intent(in) :: dicts
    type(token_t), intent(out) :: tok
    character(len=:), allocatable :: content
    integer :: start_length, end_length
    type(str_list) :: variants

    if (.not. parse_length_with_string(part, content, start_length, end_length)) then
      call make_literal_token(part, tok)
      return
    end if
    if (len(content) > 0 .and. .not. dictionaries_has(dicts, content)) then
      call make_literal_token(part, tok)
      return
    end if

    variants = dictionaries_get(dicts, content)
    ! deep copy already; ensure we own a fresh list for token_init transfer
    call token_init(tok, part, start_length, end_length, variants)
  end subroutine dictionary_tokenizer

  function unescape_commas(s) result(out)
    character(len=*), intent(in) :: s
    character(len=:), allocatable :: out
    integer :: i
    out = ''
    i = 1
    do while (i <= len(s))
      if (i < len(s) .and. s(i:i) == '\' .and. s(i+1:i+1) == ',') then
        out = out // ','
        i = i + 2
      else
        out = out // s(i:i)
        i = i + 1
      end if
    end do
  end function unescape_commas

  subroutine words_tokenizer(part, tok)
    character(len=*), intent(in) :: part
    type(token_t), intent(out) :: tok
    character(len=:), allocatable :: content, work
    integer :: start_length, end_length, idx, i
    type(str_list) :: variants
    character(len=:), allocatable :: piece

    if (.not. parse_length_with_string(part, content, start_length, end_length)) then
      call make_literal_token(part, tok)
      return
    end if

    call str_list_init(variants)
    work = content
    idx = 1
    do while (idx <= len(work))
      if (idx < len(work) .and. work(idx:idx) == '\' .and. work(idx+1:idx+1) == ',') then
        idx = idx + 2
      else if (work(idx:idx) == ',') then
        piece = work(1:idx-1)
        call str_list_push(variants, piece)
        work = work(idx+1:)
        idx = 1
      else
        idx = idx + 1
      end if
    end do
    call str_list_push(variants, work)

    do i = 1, variants%n
      variants%items(i)%s = unescape_commas(variants%items(i)%s)
    end do

    call token_init(tok, part, start_length, end_length, variants)
  end subroutine words_tokenizer

  subroutine part_to_token(part, dicts, tok)
    character(len=*), intent(in) :: part
    type(dictionaries), intent(in) :: dicts
    type(token_t), intent(out) :: tok
    character(len=1) :: first
    type(str_list) :: variants

    if (len(part) == 0) then
      call make_literal_token(part, tok)
      return
    end if

    first = part(1:1)
    select case (first)
    case ('#')
      call simple_tokenizer(part, '0123456789', tok)
    case ('@')
      call simple_tokenizer(part, 'abcdefghijklmnopqrstuvwxyz', tok)
    case ('*')
      call simple_tokenizer(part, 'abcdefghijklmnopqrstuvwxyz0123456789', tok)
    case ('-')
      call simple_tokenizer(part, &
        'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789', tok)
    case ('!')
      call simple_tokenizer(part, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', tok)
    case ('?')
      call simple_tokenizer(part, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789', tok)
    case ('&')
      call simple_tokenizer(part, &
        'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ', tok)
    case ('%')
      call dictionary_tokenizer(part, dicts, tok)
    case ('$')
      call words_tokenizer(part, tok)
    case default
      if (len(part) > 1 .and. part(1:1) == '\' .and. is_special(part(2:2))) then
        call str_list_init(variants)
        call str_list_push(variants, part(2:))
        call token_init(tok, part, 1, 1, variants)
      else
        call make_literal_token(part, tok)
      end if
    end select
  end subroutine part_to_token

  subroutine parse_pattern(input_pattern, dicts, out)
    character(len=*), intent(in) :: input_pattern
    type(dictionaries), intent(in) :: dicts
    type(token_list), intent(out) :: out
    type(str_list) :: parts
    type(token_t) :: tok
    integer :: i

    out%n = 0
    if (allocated(out%items)) deallocate(out%items)

    call split_keeping_delimiters(input_pattern, parts)
    do i = 1, parts%n
      if (len(parts%items(i)%s) == 0) cycle
      call part_to_token(parts%items(i)%s, dicts, tok)
      call token_list_push(out, tok)
    end do
    call str_list_free(parts)
  end subroutine parse_pattern

end module wl_parse_pattern
